# eBay Used Car Sales

### In this analysis we will examine data from eBay's German site regarding used car sales


This data set came from Kaggle (https://www.kaggle.com/orgesleka/used-cars-database/data) and has been sampled to reduce run time. The data comes from *eBay Kleinanzeigen*. In this project I will use NumPy and pandas to analyze used car listings.

In [1]:
import pandas as pd
import numpy as np

In [2]:
autos = pd.read_csv('autos.csv', encoding='Latin-1')

In [3]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
...                    ...                                                ...   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine_

In [4]:
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

## Cleaning the Data

So we've loaded our csv and are able to take a look at the metadata as well as a couple of our first rows. 

Our dataframe 'autos' holds 50k rows but immediately I see some of our columns are short of that mark, meaing we will need to clean this up.Columns 'vehicleType','gearbox','model','fuelType', and 'notRepairedDamage' all seem to be missing some data.

While it isn't a deal breaker, the column names are in camel case instead of Python's preferred snake case. I will make some adjustments to the column names in this regard.

Our columns look to be either an object type or integer type.

If we look at our sample set we can see a few columns that could benefit from slight modifications. The first that comes to mind is the 'odometer' column which is showing a string. It would probably be better to read that column as an integer.

In [5]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

For the sake of cleanliness, I'll edit some of these column names to be more legible as well as convert all column names from camelcase to snakecase

In [6]:
autos_col = autos.columns

In [7]:
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'num_pictures', 'postal_code',
       'last_seen']

In [8]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  num_pictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

Great. So the column titles are looking how we wanted them to look. The main reason for these changes was 1. to make sure the titles are clear and easy to read, and 2. to make the titles in the snakecase format which is preferred in Python.

What I'm going to do next is to inspect the type of data that we are working with. We might find that some of the column data is dominated by one value and can be removed from our data.

It is also likely that some columns are using strings to indicate something that would be better listed as a number. So I will make some changes to convert these columns' data.

In [9]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-05 16:57:05  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  num_pictures   postal_code            last_seen  
count                 50000       50000.0  50000.000000                50000  
unique                   76           NaN           NaN                39481  
top     2016-04-03 00:00:00           NaN           NaN  2016-04-07 06:17:27  
freq                   1946           NaN           NaN                    8  
mean                    NaN           0.0  50813.627300                  NaN  
std                     NaN           0.0  25779.747957                  NaN  
min                     NaN           0.0   1067.000000                  NaN  
25%                     NaN           0.0  30451.000000                  NaN  
50%                     NaN           0.0  49577.000000                  NaN  
75%                     NaN           0.0  71540.000000                  NaN  
max                     NaN           0.0  99998.000000                  NaN

So first I see that the price column data entries have a "$" which means that the data is stored as a string. The same goes for the odometer column and "km". So I will need to change these.

I will also change the odometer column title to "odometer_km" to ensure that the numeric value is understood to be read in km.

In [10]:
autos["price"] = autos["price"].str.replace("$","").str.replace(",","").astype(int)

In [11]:
autos.price.head()

0    5000
1    8500
2    8990
3    4350
4    1350
Name: price, dtype: int64

In [12]:
autos["odometer"] = autos["odometer"].str.replace(",","").str.replace("km","").astype(int)

In [13]:
autos.rename(columns={"odometer":"odometer_km"},inplace=True)
autos

date_crawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
...                    ...                                                ...   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine

I've cleaned up the price and odometer columns. Next I'll search for coulmns that have data that may not be useful, such as all rows having mostly the same value.

In [14]:
autos.seller.value_counts()

privat        49999
gewerblich        1
Name: seller, dtype: int64

In [15]:
autos.offer_type.value_counts()

Angebot    49999
Gesuch         1
Name: offer_type, dtype: int64

In [16]:
autos.num_pictures.value_counts()

0    50000
Name: num_pictures, dtype: int64

In [17]:
autos.unrepaired_damage.value_counts()

nein    35232
ja       4939
Name: unrepaired_damage, dtype: int64

In [18]:
del autos["seller"]
del autos["offer_type"]
del autos["num_pictures"]

In [19]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   price   abtest vehicle_type  registration_year    gearbox  power_ps  \
0   5000  control          bus               2004    manuell       158   
1   8500  control    limousine               1997  automatik       286   
2   8990     test    limousine               2009    manuell       102   
3   4350  control   kleinwagen               2007  automatik        71   
4   1350     test        kombi               2003    manuell         0   

    model  odometer_km  registration_month fuel_type       brand  \
0  andere       150000                   3       lpg     peugeot   
1     7er       150000                   6    benzin         bmw   
2    golf        70000                   7    benzin  volkswagen   
3  fortwo        70000                   6    benzin       smart   
4   focus       150000                   7    benzin        ford   

  unrepaired_damage           ad_created  postal_code            last_seen  
0              nein  2016-03-26 00:00:00        79588  2016-04-06 06:45:54  
1              nein  2016-04-04 00:00:00        71034  2016-04-06 14:45:08  
2              nein  2016-03-26 00:00:00        35394  2016-04-06 20:15:37  
3              nein  2016-03-12 00:00:00        33729  2016-03-15 03:16:28  
4              nein  2016-04-01 00:00:00        39218  2016-04-01 14:38:50

Simply glancing over some of the data, I saw that the seller, offer_type, and num_pictures columns had the same values in each row. After closer inspection, this was indeed the case. I also thought unrepaired_damage would be the same, but it turns out that there were more than enough differing values to be included in the dataframe.

The repetitive columns have been removed from the dataframe.

Next I am going to examine the price and odometer columns for possible outliers. An outlier will be any data that is irrelevant and will skew our results. 

In [20]:
autos['odometer_km'].unique().shape

(13,)

In [21]:
autos['odometer_km'].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [22]:
autos['odometer_km'].value_counts().sort_index(ascending=False)

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
40000       819
30000       789
20000       784
10000       264
5000        967
Name: odometer_km, dtype: int64

Nothing looks too out of the ordinary for the odometer column, so I don't think any changes need to be made.

In [23]:
autos['price'].unique().shape

(2357,)

In [24]:
autos['price'].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [25]:
autos['price'].value_counts().sort_index(ascending=False)

99999999       1
27322222       1
12345678       3
11111111       2
10000000       1
3890000        1
1300000        1
1234566        1
999999         2
999990         1
350000         1
345000         1
299000         1
295000         1
265000         1
259000         1
250000         1
220000         1
198000         1
197000         1
194000         1
190000         1
180000         1
175000         1
169999         1
169000         1
163991         1
163500         1
155000         1
151990         1
            ... 
66             1
65             5
60             9
59             1
55             2
50            49
49             4
47             1
45             4
40             6
35             1
30             7
29             1
25             5
20             4
18             1
17             3
15             2
14             1
13             2
12             3
11             2
10             7
9              1
8              1
5              2
3              1
2             

In [26]:
autos = autos[autos['price'].between(1,350001)]

In [27]:
autos['price'].value_counts().sort_index(ascending=False)

350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
194000      1
190000      1
180000      1
175000      1
169999      1
169000      1
163991      1
163500      1
155000      1
151990      1
145000      1
139997      1
137999      1
135000      1
130000      1
129000      1
128000      1
120000      2
119900      1
119500      1
         ... 
70         10
66          1
65          5
60          9
59          1
55          2
50         49
49          4
47          1
45          4
40          6
35          1
30          7
29          1
25          5
20          4
18          1
17          3
15          2
14          1
13          2
12          3
11          2
10          7
9           1
8           1
5           2
3           1
2           3
1         156
Name: price, Length: 2346, dtype: int64

In the price column, there were over 1000 cars that were priced at 0 Euros, and a few that were priced over half a million Euros. I cleared these out of our data to focus on more realistic listings that the average user would be looking at.

## Looking at the Data

Next up, I am going to look at some columns regarding time, specifically the date_crawled (the date the data was scraped from the site), ad_created, and last_seen. These columns are written using strings instead of datetimes, so I am only going to observe the counts for each day and see what kind of information can be drawn from them. All value counts will be shown as percentages of the total of all posts.

In [28]:
autos['date_crawled'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.025327
2016-03-06    0.014043
2016-03-07    0.036014
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-10    0.032184
2016-03-11    0.032575
2016-03-12    0.036920
2016-03-13    0.015670
2016-03-14    0.036549
2016-03-15    0.034284
2016-03-16    0.029610
2016-03-17    0.031628
2016-03-18    0.012911
2016-03-19    0.034778
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-22    0.032987
2016-03-23    0.032225
2016-03-24    0.029342
2016-03-25    0.031607
2016-03-26    0.032204
2016-03-27    0.031092
2016-03-28    0.034860
2016-03-29    0.034099
2016-03-30    0.033687
2016-03-31    0.031834
2016-04-01    0.033687
2016-04-02    0.035478
2016-04-03    0.038608
2016-04-04    0.036487
2016-04-05    0.013096
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64

In [29]:
autos['ad_created'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
2015-12-30    0.000021
2016-01-03    0.000021
2016-01-07    0.000021
2016-01-10    0.000041
2016-01-13    0.000021
2016-01-14    0.000021
2016-01-16    0.000021
2016-01-22    0.000021
2016-01-27    0.000062
2016-01-29    0.000021
2016-02-01    0.000021
2016-02-02    0.000041
2016-02-05    0.000041
2016-02-07    0.000021
2016-02-08    0.000021
2016-02-09    0.000021
2016-02-11    0.000021
2016-02-12    0.000041
2016-02-14    0.000041
2016-02-16    0.000021
2016-02-17    0.000021
2016-02-18    0.000041
2016-02-19    0.000062
2016-02-20    0.000041
2016-02-21    0.000062
                ...   
2016-03-09    0.033151
2016-03-10    0.031895
2016-03-11    0.032904
2016-03-12    0.036755
2016-03-13    0.017008
2016-03-14    0.035190
2016-03-15    0.034016
2016-03-16    0.030125
2016-03-17    0.031278
2016-03-18    0.013590
2016-03-19    0.033687
2016-03-20    0.037949
2016-03-21 

Something interesting I see in this data is that there is a definite dropoff in posts around the one month mark. This could be due to several factors. Some that comes to mind are listings expiring, users deleting and reposting listings to rise in SEO ranks, or simply because most cars are sold within a month. I checked the ebay policy regarding listing duration and these listings will be reposted after one month. That doesn't explain why there are some listings dating back ten months however.

In [30]:
autos['last_seen'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-08    0.007413
2016-03-09    0.009595
2016-03-10    0.010666
2016-03-11    0.012375
2016-03-12    0.023783
2016-03-13    0.008895
2016-03-14    0.012602
2016-03-15    0.015876
2016-03-16    0.016452
2016-03-17    0.028086
2016-03-18    0.007351
2016-03-19    0.015834
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-22    0.021373
2016-03-23    0.018532
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-26    0.016802
2016-03-27    0.015649
2016-03-28    0.020859
2016-03-29    0.022341
2016-03-30    0.024771
2016-03-31    0.023783
2016-04-01    0.022794
2016-04-02    0.024915
2016-04-03    0.025203
2016-04-04    0.024483
2016-04-05    0.124761
2016-04-06    0.221806
2016-04-07    0.131947
Name: last_seen, dtype: float64

The last_seen column keeps track of when a listing was last visited. The majority of posts have been viewed within the last four days. This is likely because users will browse several listings while looking at cars, taking into account price, odometer, age, etc. Listed cars that might have extremes in these categories, such as a high price or high odometer reading will be passed for viewing by potential buyers or simply excluded by means of search parameters.

In [31]:
autos['registration_year'].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

I wanted to take a look at the registration_year column. Interestingly, there are cars registered in the years 1000 and 9999. Clearly there has been some input errors. But it looks like the average car from the data is 12 years old (data crawled in 2016). To avoid having our data skewed by these entries, I will limit our data to exclude cars before 1950 and after 2016 (date of data crawl)

In [32]:
autos = autos[autos['registration_year'].between(1900,2017)]

In [33]:
autos['registration_year'].value_counts(normalize=True)

2000    0.065650
2005    0.061074
1999    0.060263
2004    0.056227
2003    0.056144
2006    0.055541
2001    0.054833
2002    0.051713
1998    0.049154
2007    0.047365
2008    0.046076
2009    0.043372
1997    0.040584
2011    0.033761
2010    0.033054
2017    0.028956
1996    0.028561
2012    0.027250
1995    0.025524
2016    0.025378
2013    0.016704
2014    0.013792
1994    0.013084
1993    0.008841
2015    0.008154
1992    0.007697
1990    0.007218
1991    0.007052
1989    0.003619
1988    0.002808
          ...   
1966    0.000458
1976    0.000437
1969    0.000395
1975    0.000374
1965    0.000354
1964    0.000250
1963    0.000166
1959    0.000125
1961    0.000125
1910    0.000104
1962    0.000083
1937    0.000083
1958    0.000083
1956    0.000083
1950    0.000062
1934    0.000042
1954    0.000042
1957    0.000042
1951    0.000042
1941    0.000042
1955    0.000042
1953    0.000021
1931    0.000021
1927    0.000021
1929    0.000021
1948    0.000021
1938    0.000021
1939    0.0000

Next I will take a look at the car brands in terms of frequency. I'll display brands in order of most common to least common and examine the results for any interesting information.

In [40]:
top_brands = autos['brand'].value_counts().head(10).index
autos['brand'].value_counts()

volkswagen        10220
bmw                5231
opel               5210
mercedes_benz      4610
audi               4139
ford               3357
renault            2292
peugeot            1423
fiat               1242
seat                895
skoda               777
nissan              736
mazda               733
smart               688
citroen             678
toyota              609
hyundai             480
sonstige_autos      463
volvo               437
mini                418
mitsubishi          392
honda               380
kia                 342
alfa_romeo          319
porsche             286
suzuki              282
chevrolet           271
chrysler            169
dacia               128
daihatsu            120
jeep                107
subaru              101
land_rover           99
saab                 78
daewoo               75
jaguar               73
trabant              67
rover                65
lancia               54
lada                 27
Name: brand, dtype: int64

Not surprisingly, Volkswagen is the most popular car for sale, with over 1/5th of listings being a VW. Second is BMW, with Opel being third. These three car brands together make up roughly 40% of all postings on the German ebay auto market. Ford is the most popular American car, coming in at 6th with nearly 7% of listings. Japanese cars, while popular in North America, appear in the middle, with Nissan being the most commonly listed. 

For North Americans with little interest in cars, many of these brands will be new, such as Renault, Citroen, and Peugeut from France, Skoda from Czechia, and SEAT from Spain.

At first, I did not recognize 'Sonstige Autos' between Volvo and Hyundai. I thought 'how can a brand be as popular as Volvo and Hyundai and I've never heard of it before.'  It turns out that 'sonstige' is German for 'other.' So these are all probably cars with brands that are not in the German ebay system for listing registration. However, if Lada is an available brand, I can't imagine what brands are not available.

## Finding Average Price for  Top Brands

Now that we have a good idea of what brands are available, I'm going to determine the mean price of the top brands. In the cell above, I created a list called 'top_brands'which shows the top 10 brands by number of listings.  Next, I am going to find the mean price of the listed brands and create a dictionary to display the mean prices in descending order.

In [51]:
import operator
top_cars_mean = {}

for cars in top_brands:
    car_brand = autos[autos['brand']==cars]
    mean_price = car_brand['price'].mean().round(2)
    top_cars_mean[cars] = mean_price
    
by_price = sorted(top_cars_mean.items(), key=operator.itemgetter(1), reverse=True)
by_price

[('audi', 9239.32),
 ('mercedes_benz', 8528.27),
 ('bmw', 8284.05),
 ('volkswagen', 5351.4),
 ('seat', 4354.73),
 ('ford', 3732.27),
 ('peugeot', 3069.41),
 ('opel', 2952.41),
 ('fiat', 2791.09),
 ('renault', 2442.77)]

From our top brands, we can start to distinguish groups of cars based on mean price. It's clear that Audi, BMW, and Mercedes Benz are in a class of their own. Ford, SEAT, Opel, and the rest near the bottom round out the second group. In between is Volkswagen, not really in either group. 

Being a car guy, one thing I noticed is the distinct leaning for German cars. Audi/Volkswagen, BMW, and Mercedes Benz are cleary separate from the rest. Opel is technically a German brand, but is owned by General Motors and typically produces cars on shared platforms with Chevy, Buick, and Cadillac, etc. Despite being relatively similar to Volkswagen in terms of sticker price, the resale values are much lower for American operated Opel and German operated Volkswagen.

With this fact in mind, I want to investigate the differences between the cars for sale to see if this price difference is indeed a favorability for german cars, or if it's just a coincedence. To do that, I will take a look at the average mileage of the brands.

In [55]:
mean_km = {}

for cars in top_brands:
    car_brand = autos[autos['brand']==cars]
    odometer_mean = car_brand['odometer_km'].mean().round()
    mean_km[cars] = odometer_mean
    
mean_km

{'audi': 129407.0,
 'bmw': 132667.0,
 'fiat': 117367.0,
 'ford': 124255.0,
 'mercedes_benz': 130962.0,
 'opel': 129416.0,
 'peugeot': 127316.0,
 'renault': 128096.0,
 'seat': 121626.0,
 'volkswagen': 128928.0}

In [84]:
brand_mean_km = pd.Series(mean_km)
brand_mean_price = pd.Series(top_cars_mean)
bmk_df = pd.DataFrame(brand_mean_km, columns=['mean_km'])
bmp_df = pd.DataFrame(brand_mean_price, columns=['mean_price'])
bmk_df.insert(1, 'mean_price', brand_mean_price)

bmk_df

mean_km  mean_price
audi           129407.0     9239.32
bmw            132667.0     8284.05
fiat           117367.0     2791.09
ford           124255.0     3732.27
mercedes_benz  130962.0     8528.27
opel           129416.0     2952.41
peugeot        127316.0     3069.41
renault        128096.0     2442.77
seat           121626.0     4354.73
volkswagen     128928.0     5351.40

## Conclusion

Here we can see that the German luxury brands, Audi, BMW, and Mercedes Benz, hold their value fairly well. These three brands are at the top of the list for both their mean odometer measurement as well as price, meaning sellers value these brands with more wear and tear higher than other brands of lesser wear and tear. 

Looking at Opel and Volkswagen, I think the German-owned preference can be seen here very clearly. While the average odometer reading for Opel cars is only around 500 km more, the Volkswagen average price is around 2400 euros higher than Opel. I don't think a 500 km difference can create that big of  difference in price.